code/game/machinery/PDApainter.dm

```diff
@@ -84,42 +91,96 @@
 			stat |= BROKEN
 			update_icon()
 
-/obj/machinery/pdapainter/attack_hand(mob/user as mob)
+/obj/machinery/pdapainter/attack_hand(mob/user)
 	if(..())
-		return 1
+		return TRUE
 
-	src.add_fingerprint(user)
+	ui_interact(user)
 
-	if(storedpda)
-		var/obj/item/pda/P
-		P = input(user, "Select your color!", "PDA Painting") as null|anything in colorlist
-		if(!P)
-			return
-		if(!in_range(src, user))
-			return
+/obj/machinery/pdapainter/ui_interact(mob/user, ui_key, datum/tgui/ui, force_open, datum/tgui/master_ui, datum/ui_state/state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "PdaPainter", "PDA Painter", 510, 510)
+		ui.open()
 
-		storedpda.icon_state = P
-		storedpda.desc = colorlist[P]
+/obj/machinery/pdapainter/ui_data(mob/user)
+	var/list/data = list()
+	data["has_pda"] = storedpda ? TRUE : FALSE
+	if(storedpda)
+		data["current_appearance"] = cached_icon_state
 
-	else
-		to_chat(user, "<span class='notice'>[src] is empty.</span>")
+		var/icon/preview_sprite = icon(storedpda.icon, preview_icon_state, frame = 1)
+		data["preview_appearance"] = icon2base64(preview_sprite)
 
+	return data
 
-/obj/machinery/pdapainter/verb/ejectpda()
-	set name = "Eject PDA"
-	set category = "Object"
-	set src in oview(1)
+/obj/machinery/pdapainter/ui_static_data(mob/user)
+	var/list/data = list()
+	data["pda_colors"] = colorlist
+	return data
 
-	if(usr.incapacitated())
+/obj/machinery/pdapainter/ui_act(action, params, datum/tgui/ui)
+	if(..())
 		return
 
+	. = TRUE
+
+	switch(action)
+		if("insert_pda")
+			insertpda(ui.user)
+		if("eject_pda")
+			ejectpda(ui.user)
+		if("choose_pda")
+			preview_icon_state = params["selectedPda"]
+		if("paint_pda")
+			paintpda()
+
+	if(.)
+		add_fingerprint(ui.user)
+
+/obj/machinery/pdapainter/proc/insertpda(mob/user)
 	if(storedpda)
-		storedpda.loc = get_turf(src.loc)
-		storedpda = null
-		update_icon()
-	else
+		to_chat(user, "There is already a PDA inside.")
+		return
+	if(!ishuman(user))
+		return
+
+	var/obj/item/pda/P = user.get_active_hand()
+
+	if(istype(P))
+		if(user.drop_item())
+			storedpda = P
```

Comment:
I want to merge this, but this could cause potential GC issues, could you register signals on this and call `ejectpda` if this gets qdel'd for some reason?

Reply:
I've added on_pda_qdel, which is called when the PDA gets deleted while inside the PDA Painter. I think this is what you wanted but I'm not sure